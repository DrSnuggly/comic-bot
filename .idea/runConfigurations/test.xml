<component name="ProjectRunConfigurationManager">
  <configuration default="false" name="test" type="CompoundRunConfigurationType">
    <toRun name="debugger" type="ChromiumRemoteDebugType" />
    <toRun name="All Tests" type="JavaScriptTestRunnerVitest" />
    <method v="2" />
  </configuration>
</component>