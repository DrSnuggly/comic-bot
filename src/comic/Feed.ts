import type { Feed as RssFeed, FeedItem as RssFeedItem } from "domutils"
import { decodeXML } from "entities"
import { parseFeed } from "htmlparser2"
import type { ComicData } from "../schema"
import { ComicError } from "./ComicError"

export type FeedData = {
  feedName: string
  dateUpdated: Date
  pageName: string
  pageLink: string
}

export class FeedError extends ComicError {}

export class Feed implements FeedData {
  /**
   * @param comic - the target comic.
   * @param content - the RSS feed content as an object.
   * @protected
   */
  protected constructor(
    public comic: ComicData,
    public content: RssFeed,
  ) {}

  /**
   * Retrieve the feed content.
   * @param comic - the target comic.
   * @throws FeedError if feed cannot be retrieved.
   */
  static async fromComic(comic: ComicData): Promise<Feed> {
    const response = await fetch(comic.rssUrl)
    return Feed.fromResponse(response, comic)
  }

  /**
   * Retrieve the feed content.
   * @param response - the response to parse.
   * @param comic - the target comic.
   * @throws FeedError if feed cannot be parsed.
   */
  static async fromResponse(
    response: Response,
    comic: ComicData,
  ): Promise<Feed> {
    if (!response.ok) {
      throw new FeedError(
        comic,
        `received ${response.status} when retrieving ${response.url}: ${await response.text()}`,
      )
    }

    const content = parseFeed(await response.text())
    if (content === null) {
      throw new FeedError(comic, "could not parse feed")
    }
    return new Feed(comic, content)
  }

  static formatDateAsName(date: Date): string {
    return date.toISOString().substring(0, 10)
  }

  /**
   * Get the RSS feed title.
   */
  get feedName(): string {
    if (this.content.title === undefined) {
      return this.comic.name
    }
    return decodeXML(this.content.title)
  }

  #latest: RssFeedItem | undefined
  /**
   * Get the latest item in the RSS feed.
   * @throws FeedError if no items exist in the feed.
   */
  get latest(): RssFeedItem {
    if (this.#latest === undefined) {
      const latest = this.content.items[0]
      if (latest === undefined) {
        throw new FeedError(this.comic, "no items in feed")
      }
      this.#latest = latest
    }
    return this.#latest
  }

  #dateUpdated: Date | undefined
  /**
   * Get when the feed was last updated.
   * @throws FeedError if last update date cannot be determined.
   */
  get dateUpdated(): Date {
    if (this.#dateUpdated === undefined) {
      if (this.latest.pubDate === undefined) {
        if (this.content.updated === undefined) {
          throw new FeedError(
            this.comic,
            "could not determine when last updated",
          )
        }
        this.#dateUpdated = this.content.updated
      } else {
        this.#dateUpdated = this.latest.pubDate
      }
    }
    return this.#dateUpdated
  }

  /**
   * Get the title for the latest item, falling back to the publishing date.
   */
  get pageName(): string {
    if (this.latest.title) {
      return decodeXML(this.latest.title)
    }
    return Feed.formatDateAsName(this.dateUpdated)
  }

  /**
   * Get the link to the latest page.
   * @throws FeedError if no link is found.
   */
  get pageLink(): string {
    if (!this.latest.link) {
      throw new FeedError(this.comic, "latest page URL missing")
    }
    return this.latest.link
  }
}
