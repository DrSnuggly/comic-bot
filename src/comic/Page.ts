import { decodeHTML } from "entities"
import type { ComicData } from "../schema"
import { ComicError } from "./ComicError"

export type PageData = {
  imageUri: string
  altText?: string | undefined
}

export class PageError extends ComicError {}

/**
 * Content handler for {@link HTMLRewriter}.
 */
export class Page implements PageData {
  imageUri: string
  altText: string | undefined

  /**
   * @param comic - the target comic.
   * @param rewriter - the instantiated {@link HTMLRewriter} object.
   * @param data - the parsed page data.
   */
  protected constructor(
    public comic: ComicData,
    protected rewriter: HTMLRewriter,
    data: PageData,
  ) {
    this.imageUri = data.imageUri
    this.altText = data.altText
  }

  /**
   * Fetch and parse the comic.
   * @param url - the target comic page.
   * @param comic - the target comic.
   * @param rewriter - the instantiated {@link HTMLRewriter} object.
   */
  static async fromUrl(
    url: string,
    comic: ComicData,
    rewriter: HTMLRewriter,
  ): Promise<Page> {
    const response = await fetch(url)
    return Page.fromResponse(response, comic, rewriter)
  }

  /**
   * Fetch and parse the comic.
   * @param response - the target comic page response.
   * @param comic - the target comic.
   * @param rewriter - the instantiated {@link HTMLRewriter} object.
   * @throws PageError if bad response or unable to parse.
   */
  static async fromResponse(
    response: Response,
    comic: ComicData,
    rewriter: HTMLRewriter,
  ): Promise<Page> {
    if (!response.ok) {
      throw new PageError(
        comic,
        `received ${response.status} when retrieving ${response.url}: ${await response.text()}`,
      )
    }

    let imageUri: string | undefined
    let altText: string | undefined
    const handler: HTMLRewriterElementContentHandlers = {
      element(element) {
        // Get comic image, if not already found.
        if (imageUri === undefined) {
          const src = element.getAttribute("src")
          if (src) imageUri = src
        }

        // Get comic alt text, if not already found.
        if (altText === undefined) {
          const title = Page.normalizeAltText(element.getAttribute("title"))
          if (title !== undefined) altText = title
        }
      },
      text(element) {
        // Get comic alt text, if not already found.
        if (altText === undefined) {
          const text = Page.normalizeAltText(element.text)
          if (text !== undefined) altText = text
        }
      },
    }

    let parser = rewriter.on(comic.imageSelector, handler)
    if (comic.altSelector) {
      parser = parser.on(comic.altSelector, handler)
    }
    await parser.transform(response).text()
    if (imageUri === undefined) {
      throw new PageError(comic, "could not find image URI")
    }

    return new Page(comic, rewriter, { imageUri, altText })
  }

  /**
   * Normalize input into the expected value.
   * @param input - the input to normalize.
   */
  static normalizeAltText(input: unknown): string | undefined {
    if (!input || typeof input !== "string") return undefined
    const parsed = decodeHTML(input).trim()
    if (!parsed) return undefined
    return parsed
  }
}
