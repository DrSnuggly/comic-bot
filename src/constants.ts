export const KV_INDEX_KEY = "index"
