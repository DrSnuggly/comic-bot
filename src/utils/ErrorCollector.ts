/**
 * Guard for non-string iterables.
 * @param obj - the item to test.
 */
function isNonStringIterable(
  obj: unknown,
): obj is Exclude<Iterable<unknown>, string> {
  if (typeof obj === "string") return false
  return typeof (obj as Iterable<unknown>)[Symbol.iterator] === "function"
}

export class ErrorCollector extends Set<unknown> {
  override add(e: unknown): this {
    if (!isNonStringIterable(e)) return super.add(e)
    for (const item of e) super.add(item)
    return this
  }

  /**
   * Add any errors from `await Promise.allSettled()`.
   * @param settled - the settled promises array.
   */
  addSettled(settled: PromiseSettledResult<unknown>[]): this {
    for (const item of settled.filter((item) => item.status === "rejected")) {
      this.add(item.reason)
    }
    return this
  }

  /**
   * Assert that no items have been collected.
   * @throws unknown[] for the collected errors.
   */
  assertEmpty(): void {
    if (this.size) throw Array.from(this)
  }
}
