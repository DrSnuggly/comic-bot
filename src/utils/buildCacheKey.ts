import type { ComicData } from "../schema"

/**
 * Generate a cache key for storing and retrieving last update times.
 * @param url - a webhook URL.
 * @param comic - the target comic.
 */
export function buildCacheKey(url: string, comic: ComicData): string {
  return `${comic.rssUrl}|${url}`
}
