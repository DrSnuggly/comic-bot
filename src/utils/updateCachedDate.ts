import { env } from "cloudflare:workers"
import type { ComicData } from "../schema"
import { buildCacheKey } from "./buildCacheKey"

/**
 * Update the cached date for a given comic and webhook.
 * @param url - a webhook URL.
 * @param comic - the target comic.
 */
export function updateCachedDate(url: string, comic: ComicData): Promise<void> {
  return env.KV.put(buildCacheKey(url, comic), new Date().toISOString())
}
