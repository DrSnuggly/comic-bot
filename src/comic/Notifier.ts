import { env } from "cloudflare:workers"
import type {
  APIEmbed,
  RESTPostAPIWebhookWithTokenJSONBody,
} from "discord-api-types/v10"
import { type ComicData, lastUpdateMapSchema } from "../schema"
import { buildCacheKey } from "../utils/buildCacheKey"
import { ErrorCollector } from "../utils/ErrorCollector"
import { updateCachedDate } from "../utils/updateCachedDate"
import { ComicError } from "./ComicError"
import type { FeedData } from "./Feed"
import type { PageData } from "./Page"

export type Data = {
  feed: FeedData
  page: PageData
}

export class NotifierError extends ComicError {}

export class Notifier {
  /**
   * @param targets - the webhooks to update.
   * @param feed - the feed data to use.
   * @param page - the page data to use.
   * @param comic - the target comic.
   */
  constructor(
    public targets: string[],
    public feed: FeedData,
    public page: PageData,
    public comic: ComicData,
  ) {}

  /**
   * Determine targets from cache.
   * @param data - the data to use for generating embeds.
   * @param comic - the target comic.
   */
  static async fromData(data: Data, comic: ComicData) {
    if (!comic.webhooks.length) {
      throw new NotifierError(comic, "no webhook targets configured")
    }

    // Get webhooks that still need updating.
    const kvKeys = comic.webhooks.map((url) => buildCacheKey(url, comic))
    const updateMap = lastUpdateMapSchema.parse(await env.KV.get(kvKeys))
    const targets = comic.webhooks.filter((url) => {
      const webhookUpdated = updateMap.get(buildCacheKey(url, comic))
      if (!webhookUpdated) return true
      return webhookUpdated < data.feed.dateUpdated
    })

    return new Notifier(targets, data.feed, data.page, comic)
  }

  protected updateCachedDate(url: string): Promise<void> {
    return updateCachedDate(url, this.comic)
  }

  /**
   * Notify the configured Discord webhook URLs of any updates, if needed.
   */
  async send() {
    if (!this.targets.length) {
      console.log(new NotifierError(this.comic, "no webhooks to send").message)
      return
    }
    await this.sendUpdates(this.embeds)
  }

  /**
   * Generate Discord webhook embeds.
   */
  get embeds(): APIEmbed[] {
    return [
      {
        title: this.feed.pageName,
        author: { name: this.feed.feedName },
        image: { url: this.page.imageUri },
        url: this.feed.pageLink,
      },
      // Only include alt-text embed if there's alt-text.
      ...(this.page.altText
        ? [{ description: `Alt text: ||${this.page.altText}||` }]
        : []),
    ]
  }

  /**
   * Send updates to the given targets with the given embeds.
   * @param embeds - the embeds to send.
   * @protected
   */
  protected async sendUpdates(embeds: APIEmbed[]) {
    const errors = new ErrorCollector()
    const results = await Promise.allSettled(
      this.targets.map(async (url) => this.sendUpdate(url, embeds)),
    )
    // Re-throw any errors.
    errors.addSettled(results)
    errors.assertEmpty()
  }

  /**
   * Send updates to the given target with the given embeds.
   * @param url - Discord webhook URL to send to.
   * @param embeds - the embeds to send.
   * @throws NotifierError if posting to webhook failed.
   * @protected
   */
  protected async sendUpdate(url: string, embeds: APIEmbed[]) {
    const response = await fetch(url, {
      method: "post",
      body: JSON.stringify({
        embeds,
        username: env.DISCORD_USERNAME,
        ...(env.DISCORD_AVATAR_URL
          ? { avatarURL: env.DISCORD_AVATAR_URL }
          : {}),
      } as RESTPostAPIWebhookWithTokenJSONBody),
    })
    if (!response.ok) {
      throw new NotifierError(
        this.comic,
        `failed posting to ${url} with ${response.status}: ${await response.text()}`,
      )
    }

    await this.updateCachedDate(url)
  }
}
