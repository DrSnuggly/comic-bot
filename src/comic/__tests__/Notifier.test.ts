import { fetchMock, type Interceptable } from "cloudflare:test"
import { comicData, origin } from "@tests/utils/constants"
import { fetchAsset } from "@tests/utils/fetchAsset"
import { getCachedDates } from "@tests/utils/getCachedDates"
import { updateCachedDate } from "../../utils/updateCachedDate"
import { Feed, type FeedData } from "../Feed"
import { type Data, Notifier, NotifierError } from "../Notifier"
import { Page, type PageData } from "../Page"

const nowDate = new Date("2025-09-15T00:00:00Z")
const hourAgoDate = new Date("2025-09-14T23:00:00Z")
const weekAgoDate = new Date("2025-09-08T00:00:00Z")

let mockPool: Interceptable
let rewriter: HTMLRewriter
let feed: FeedData
let page: PageData
let data: Data
beforeAll(async () => {
  rewriter = new HTMLRewriter()
  fetchMock.activate()
  fetchMock.disableNetConnect()
  mockPool = fetchMock.get(origin)
  feed = await Feed.fromResponse(
    await fetchAsset("comic-feeds/standard.xml"),
    comicData,
  )
  page = await Page.fromResponse(
    await fetchAsset("comic-pages/inline-alt-text.html"),
    comicData,
    rewriter,
  )
  data = { feed, page }
})
beforeEach(() => {
  vi.useFakeTimers()
  vi.setSystemTime(nowDate)
})
afterEach(() => {
  // Config
  vi.useRealTimers()
  // Assertions
  fetchMock.assertNoPendingInterceptors()
})

describe("targeting", () => {
  it("should reject if no configured webhooks", async () => {
    const testComic = { ...comicData, webhookUrls: [] }

    const notifier = Notifier.fromData(data, testComic)

    await expect(notifier).rejects.instanceof(NotifierError)
  })

  it("should target all webhooks on first run", async () => {
    const targets = Notifier.fromData(data, comicData).then(
      (notifier) => notifier.targets,
    )

    await expect(targets).resolves.toEqual(comicData.webhookUrls)
  })

  it("should target all webhooks if all outdated", async () => {
    vi.setSystemTime(weekAgoDate)
    await Promise.all(
      comicData.webhookUrls.map((webhook) =>
        updateCachedDate(webhook, comicData),
      ),
    )
    vi.setSystemTime(nowDate)
    const targets = Notifier.fromData(data, comicData).then(
      (notifier) => notifier.targets,
    )

    await expect(targets).resolves.toEqual(comicData.webhookUrls)
  })

  it("should target only outdated webhooks", async () => {
    const toTarget = [comicData.webhookUrls[0], comicData.webhookUrls[2]]
    const toIgnore = [comicData.webhookUrls[1]]

    vi.setSystemTime(weekAgoDate)
    await Promise.all(
      toTarget.map((webhook) => updateCachedDate(webhook, comicData)),
    )
    vi.setSystemTime(hourAgoDate)
    await Promise.all(
      toIgnore.map((webhook) => updateCachedDate(webhook, comicData)),
    )
    vi.setSystemTime(nowDate)
    const targets = Notifier.fromData(data, comicData).then(
      (notifier) => notifier.targets,
    )

    await expect(targets).resolves.toEqual(toTarget)
  })

  it("should target nothing if all up-to-date", async () => {
    vi.setSystemTime(hourAgoDate)
    await Promise.all(
      comicData.webhookUrls.map((webhook) =>
        updateCachedDate(webhook, comicData),
      ),
    )
    vi.setSystemTime(nowDate)
    const targets = Notifier.fromData(data, comicData).then(
      (notifier) => notifier.targets,
    )

    await expect(targets).resolves.toEqual([])
  })
})

describe("embeds", () => {
  it("should include alt-text embed if present", () => {
    const notifier = new Notifier(comicData.webhookUrls, feed, page, comicData)

    expect(notifier.embeds).toMatchSnapshot()
  })

  it("should not include alt-text embed if missing", () => {
    const testPage: PageData = { imageUri: page.imageUri }
    const notifier = new Notifier(
      comicData.webhookUrls,
      feed,
      testPage,
      comicData,
    )

    expect(notifier.embeds).toMatchSnapshot()
  })
})

describe("sending", () => {
  it("should send embeds as body", async () => {
    const notifier = new Notifier(comicData.webhookUrls, feed, page, comicData)
    const embeds = notifier.embeds

    for (const webhook of comicData.webhookUrls) {
      mockPool
        .intercept({
          path: webhook,
          method: "post",
          body: (body) => {
            expect(JSON.parse(body)).toMatchObject({ embeds })
            return true
          },
        })
        .reply(200)
    }

    await expect(notifier.send()).resolves.toBe(undefined)
  })

  it("should only send to targets", async () => {
    const toTarget = [comicData.webhookUrls[0], comicData.webhookUrls[2]]
    const notifier = new Notifier(toTarget, feed, page, comicData)

    for (const webhook of toTarget) {
      mockPool.intercept({ path: webhook, method: "post" }).reply(200)
    }

    await expect(notifier.send()).resolves.toBe(undefined)
  })

  it("should update cached dates", async () => {
    const notifier = new Notifier(comicData.webhookUrls, feed, page, comicData)

    for (const webhook of comicData.webhookUrls) {
      mockPool.intercept({ path: webhook, method: "post" }).reply(200)
    }

    await expect(notifier.send()).resolves.toBe(undefined)
    const cache = await getCachedDates()
    expect(cache.size).toBe(comicData.webhookUrls.length)
    for (const value of cache.values()) {
      expect(value).toEqual(nowDate)
    }
  })

  it("should re-throw request failures", async () => {
    const notifier = new Notifier(comicData.webhookUrls, feed, page, comicData)

    for (const webhook of comicData.webhookUrls) {
      mockPool.intercept({ path: webhook, method: "post" }).reply(400)
    }

    await expect(notifier.send()).rejects.toMatchObject(
      comicData.webhookUrls.map(() => expect.any(NotifierError)),
    )
  })

  it("should only update cached dates if request is ok", async () => {
    const toSucceed = [comicData.webhookUrls[0], comicData.webhookUrls[2]]
    const toFail = [comicData.webhookUrls[1]]
    const notifier = new Notifier(comicData.webhookUrls, feed, page, comicData)

    for (const webhook of toSucceed) {
      mockPool.intercept({ path: webhook, method: "post" }).reply(200)
    }
    for (const webhook of toFail) {
      mockPool.intercept({ path: webhook, method: "post" }).reply(400)
    }

    await expect(notifier.send()).rejects.toMatchObject([
      expect.any(NotifierError),
    ])
    const cache = await getCachedDates()
    expect(cache.size).toBe(toSucceed.length)
    for (const value of cache.values()) {
      expect(value).toEqual(nowDate)
    }
  })

  it("should do nothing if no targets", async () => {
    const notifier = new Notifier([], feed, page, comicData)
    const consoleSpy = vi.spyOn(console, "log").mockImplementationOnce(() => {})

    await expect(notifier.send()).resolves.toBe(undefined)
    expect(consoleSpy).toHaveBeenCalled()
  })
})
