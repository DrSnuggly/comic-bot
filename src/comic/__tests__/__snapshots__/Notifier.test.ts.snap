// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`embeds > should include alt-text embed if present 1`] = `
[
  {
    "author": {
      "name": "Comic Title",
    },
    "image": {
      "url": "https://localhost/image.png",
    },
    "title": "Comic 1",
    "url": "https://localhost/comic",
  },
  {
    "description": "Alt text: ||Alt text.||",
  },
]
`;

exports[`embeds > should not include alt-text embed if missing 1`] = `
[
  {
    "author": {
      "name": "Comic Title",
    },
    "image": {
      "url": "https://localhost/image.png",
    },
    "title": "Comic 1",
    "url": "https://localhost/comic",
  },
]
`;
